lit_tests/kernel/wave/promotion.py

```diff
@@ -0,0 +1,178 @@
+    constraints: list[tkw.Constraint] = [tkw.WorkgroupConstraint(M, BLOCK_M, 0)]
+    constraints += [tkw.WorkgroupConstraint(N, BLOCK_N, 1)]
+    constraints += [
+        tkw.HardwareConstraint(
+            threads_per_wave=64,
+            waves_per_block=(1, 1, 1),
+            mma_type=tkw.MMAType.F32_16x16x16_F16,
+        )
+    ]
+
+    with tk.gen.TestLaunchContext(
+        {
+            BLOCK_M: 32,
+            BLOCK_N: 32,
+            ADDRESS_SPACE: GLOBAL_ADDRESS_SPACE,
+        }
+    ):
+        trace: CapturedTrace = read_write_same_size()
+        graph: fx.Graph = trace.get_root_graph()
+        read_node = get_read_nodes(graph)[0]
+        IndexingContext.current().finalize()
+        promote_node(read_node, SHARED_ADDRESS_SPACE)
+        print_trace(trace)
+        # CHECK: %a
+        # CHECK-NEXT: %c
+        # CHECK-NEXT: %read
+        # CHECK-SAME: (%a, 4)
+        # CHECK-NEXT: %allocate
+        # CHECK-SAME: ((M, N), f16, SHARED_ADDRESS_SPACE)
+        # CHECK-NEXT: %write_1
+        # CHECK-SAME: (%read, %allocate, 4)
+        # CHECK-NEXT: %read_1
+        # CHECK-SAME: (%allocate, 4)
+        # CHECK-NEXT: %write
+        # CHECK-SAME: (%read_1, %c, 4)
+
+        # CHECK: -----
+
+
+@tkw.wave_trace_only()
+def gemm(
+    a: tkl.Memory[M, K, ADDRESS_SPACE, tkl.f16],
+    b: tkl.Memory[N, K, ADDRESS_SPACE, tkl.f16],
+    c: tkl.Memory[M, N, ADDRESS_SPACE, tkl.f32],
+):
+    c_reg = tkl.Register[M, N, tkl.f32](0.0)
+
+    @tkw.reduction(K, init_args=[c_reg])
+    def repeat(acc: tkl.Register[M, N, tkl.f32]) -> tkl.Register[M, N, tkl.f32]:
+        a_reg = tkw.read(a, elements_per_thread=4)
+        b_reg = tkw.read(b, elements_per_thread=4)
+        acc = tkw.mma(a_reg, b_reg, acc)
+        return acc
+
+    tkw.write(repeat, c, elements_per_thread=4)
+
+
+@run
+def test_gemm():
+    constraints: list[tkw.Constraint] = [tkw.WorkgroupConstraint(M, BLOCK_M, 0)]
+    constraints += [tkw.WorkgroupConstraint(N, BLOCK_N, 1)]
+    constraints += [tkw.WorkgroupConstraint(K, BLOCK_K, 2)]
+    constraints += [
+        tkw.HardwareConstraint(threads_per_wave=64, waves_per_block=(1, 1, 1))
+    ]
+    with tk.gen.TestLaunchContext(
+        {
+            BLOCK_M: 32,
+            BLOCK_N: 32,
+            BLOCK_K: 32,
+        }
+    ):
+        trace: CapturedTrace = gemm()
+        graph: fx.Graph = trace.get_subgraph("region_0")
+        read_nodes = get_read_nodes(graph)
+        for read_node in read_nodes:
+            promote_node(read_node, SHARED_ADDRESS_SPACE)
+        hoist_allocs(trace)
+        IndexingContext.current().finalize()
+        print_trace(trace)
+        # Root graph:
+        # CHECK: %a
+        # CHECK-NEXT: %b
+        # CHECK-NEXT: %c
+        # CHECK-NEXT: %register
+        # CHECK-NEXT: %allocate
+        # CHECK-SAME: ((M, K), f16, SHARED_ADDRESS_SPACE)
+        # CHECK-NEXT: %allocate_1
+        # CHECK-SAME: ((N, K), f16, SHARED_ADDRESS_SPACE)
+        # CHECK-NEXT: reduction
+
```

Comment:
After the reduction I have an additional `%write(%reduction, %c, 4)`
If that is expected we should add:
```suggestion
        # CHECK-SAME: ((N, K), f16, SHARED_ADDRESS_SPACE)
        # CHECK-NEXT: reduction
        # CHECK-NEXT: %write
        # CHECK-SAME: (%reduction, %c, 4)

```

Reply:
Yes that is expected, will add that in. Thanks!